src/allreduce_robust.cc

```diff
@@ -102,13 +162,7 @@ void AllreduceRobust::Allreduce(void *sendrecvbuf_,
     if (prepare_fun != NULL) prepare_fun(prepare_arg);
     return;
   }
-  bool recovered = RecoverExec(sendrecvbuf_, type_nbytes * count, 0, seq_counter);
-  // now we are free to remove the last result, if any
-  if (resbuf.LastSeqNo() != -1 &&
-      (result_buffer_round == -1 ||
-       resbuf.LastSeqNo() % result_buffer_round != rank % result_buffer_round)) {
-    resbuf.DropLast();
```

Comment:
this is actually relevant to my confusion, looks like you keep the newly added cached content (like a sketch) in memory forever, otherwise it is impossible to recover after 100s of iterations later...

I am very curious how you do that as the seqno is reused per iteration, is it achieved here? what's the impact of not cleaning the buffer after recovery?

Reply:
It's old code which I haven't changed.  In short, anything relate to resbuf were not changed.  From what I read, resbuf is the storage section where it might store more than seq_counter linked allreduce/broadcast payload. 

The original rabit recovery beyond loadcheckpoint relay on seq_counter alignment. a quick example

worker 0 allreduce -> do sorting -> broadcast (seq_counter = 1, paused because link to worker 1 reset) -> consensus process involving worker 0 and worker 1 -> checkpoint
worker 1 allreduce -> do sorting -> xxxx(dead) -> loadcheckpoint -> allreduce(call RecoverExec with seq_counter = 0) -> consensus process involving worker 0 and worker 1 -> allreduce with copied seq0 allreduce result from worker 0 (detail here https://github.com/dmlc/rabit/pull/98/files#diff-6abf0a70e514e75271b0281b21669e1bR1049) -> do sorting -> broadcast(seq_counter = 1) -> consensus process involving worker 0 and 1 (no seq number difference hence go here https://github.com/dmlc/rabit/pull/98/files#diff-6abf0a70e514e75271b0281b21669e1bR1052) -> fallback to tryallreduce with exactly same boradcast calls in both worker 0 and worker 1) -> from this point, it's fully synced.


https://github.com/dmlc/rabit/pull/98/files#diff-6abf0a70e514e75271b0281b21669e1bR1025

Reply:
this part of code was removed from the PR, I was asking why removed?

Reply:
seq_diff flag is used to mark if every node has same seq_counter, if so, it means all nodes are synced and no need to recover from resbuf in other nodes.

local and global minimal seq_counter is key mono-incremental counter used to align failed worker recovery. if local seq_counter is global one, it means either local is falling behind and get data from other nodes.

Reply:
Are you answering my question about “why removed?”

Reply:
> Are you answering my question about “why removed?”
I guess I removed it in previous commiets, I don't see why broadcast needs to remove last resbuf at first place v.s allreduce not. I can add em back.

Reply:
if you don't understand why, could you either figure it out or keep it there?

Reply:
fixed, adding back both broadcast and allreduce